js/misc/paygManager.js

```diff
@@ -0,0 +1,410 @@
+    60 * 60 * 24, // 1 day
+    60 * 60 * 2,  // 2 hours
+    60 * 60,      // 1 hour
+    60 * 30,      // 30 minutes
+    60 * 2,       // 2 minutes
+    30,           // 30 seconds
+];
+
+// Takes an UNIX timestamp (in seconds) and returns a string
+// with a precision level appropriate to show to the user.
+//
+// The returned string will be formatted just in seconds for times
+// under 1 minute, in minutes for times under 2 hours, in hours and
+// minutes (if applicable) for times under 1 day, and then in days
+// and hours (if applicable) for anything longer than that in days.
+//
+// Some examples:
+//   - 45 seconds => "45 seconds"
+//   - 60 seconds => "1 minute"
+//   - 95 seconds => "1 minute"
+//   - 120 seconds => "2 minutes"
+//   - 3600 seconds => "60 minutes"
+//   - 4500 seconds => "75 minutes"
+//   - 7200 seconds => "2 hours"
+//   - 8640 seconds => "2 hours 24 minutes"
+//   - 86400 seconds => "1 day"
+//   - 115200 seconds => "1 day 8 hours"
+//   - 172800 seconds => "2 days"
+function timeToString(seconds) {
+    if (seconds < 60)
+        return Gettext.ngettext('%s second', '%s seconds', seconds).format(Math.floor(seconds));
+
+    let minutes = Math.floor(seconds / 60);
+    if (minutes < 120)
+        return Gettext.ngettext('%s minute', '%s minutes', minutes).format(minutes);
+
+    let hours = Math.floor(minutes / 60);
+    if (hours < 24) {
+        let hoursStr = Gettext.ngettext('%s hour', '%s hours', hours).format(hours);
+
+        let minutesPast = minutes % 60;
+        if (minutesPast == 0)
+            return hoursStr;
+
+        let minutesStr = Gettext.ngettext('%s minute', '%s minutes', minutesPast).format(minutesPast);
+        return ("%s %s").format(hoursStr, minutesStr);
+    }
+
+    let days = Math.floor(hours / 24);
+    let daysStr = Gettext.ngettext('%s day', '%s days', days).format(days);
+
+    let hoursPast = hours % 24;
+    if (hoursPast == 0)
+        return daysStr;
+
+    let hoursStr = Gettext.ngettext('%s hour', '%s hours', hoursPast).format(hoursPast);
+    return ("%s %s").format(daysStr, hoursStr);
+}
+
+var PaygManager = GObject.registerClass({
+    Signals: {
+        'code-expired': {},
+        'code-format-changed': {},
+        'enabled-changed': { param_types: [GObject.TYPE_BOOLEAN] },
+        'expiry-time-changed': { param_types: [GObject.TYPE_INT64] },
+        'initialized': {},
+    },
+}, class PaygManager extends GObject.Object {
+
+    _init() {
+        super._init();
+
+        this._initialized = false;
+        this._proxy = null;
+
+        this._enabled = false;
+        this._expiryTime = 0;
+        this._rateLimitEndTime = 0;
+        this._codeFormat = '';
+        this._codeFormatRegex = null;
+        this._notification = null;
+
+        // Keep track of clock changes to update notifications.
+        this._wallClock = new GnomeDesktop.WallClock({ time_only: true });
+        this._wallClock.connect('notify::clock', this._clockUpdated.bind(this));
```

Comment:
I think we can get rid of this. The expiration notifications give the time remaining, not a wall clock time, and we already have callbacks for when the expiry time changes.

Reply:
I'm not confident enough to drop this code and check that it works with the current limitations of time :/